exchange/price_granularity_test.go

```diff
@@ -28,32 +31,111 @@ func TestGetPriceBucketString(t *testing.T) {
 		},
 	}
 
-	price := 1.87
-	getOnePriceBucket(t, "low", low, price, "1.50")
-	getOnePriceBucket(t, "medium", medium, price, "1.80")
-	getOnePriceBucket(t, "high", high, price, "1.87")
-	getOnePriceBucket(t, "auto", auto, price, "1.85")
-	getOnePriceBucket(t, "dense", dense, price, "1.87")
-	getOnePriceBucket(t, "custom1", custom1, price, "1.86")
-
-	// test a cpm above the max in low price bucket
-	price = 5.72
-	getOnePriceBucket(t, "low", low, price, "5.00")
-	getOnePriceBucket(t, "medium", medium, price, "5.70")
-	getOnePriceBucket(t, "high", high, price, "5.72")
-	getOnePriceBucket(t, "auto", auto, price, "5.70")
-	getOnePriceBucket(t, "dense", dense, price, "5.70")
-	getOnePriceBucket(t, "custom1", custom1, price, "5.70")
-
-}
-
-func getOnePriceBucket(t *testing.T, name string, granularity openrtb_ext.PriceGranularity, price float64, expected string) {
-	t.Helper()
-	priceBucket, err := GetCpmStringValue(price, granularity)
-	if err != nil {
-		t.Errorf("Granularity: %s :: GetPriceBucketString: %s", name, err.Error())
+	// Define test cases
+	type aTest struct {
+		granularityId       string
+		inGranularity       openrtb_ext.PriceGranularity
+		expectedPriceBucket string
+		expectedError       error
+	}
+	testGroups := []struct {
+		groupDesc string
+		inCpm     float64
+		testCases []aTest
+	}{
+		{
+			"cpm below the max in every price bucket",
+			1.87,
+			[]aTest{
+				{"low", low, "1.50", nil},
+				{"medium", medium, "1.80", nil},
+				{"high", high, "1.87", nil},
+				{"auto", auto, "1.85", nil},
+				{"dense", dense, "1.87", nil},
+				{"custom1", custom1, "1.86", nil},
+			},
+		},
+		{
+			"cpm above the max in low price bucket",
+			5.72,
+			[]aTest{
+				{"low", low, "5.00", nil},
+				{"medium", medium, "5.70", nil},
+				{"high", high, "5.72", nil},
+				{"auto", auto, "5.70", nil},
+				{"dense", dense, "5.70", nil},
+				{"custom1", custom1, "5.70", nil},
+			},
+		},
+		{
+			"Precision value corner cases",
+			1.876,
+			[]aTest{
+				{
+					"Negative precision defaults to number of digits already in CPM float",
+					openrtb_ext.PriceGranularity{Precision: -1, Ranges: []openrtb_ext.GranularityRange{{Max: 5, Increment: 0.05}}},
+					"1.85",
+					nil,
+				},
+				{
+					"Precision value equals zero, we expect to round up to the nearest integer",
+					openrtb_ext.PriceGranularity{Ranges: []openrtb_ext.GranularityRange{{Max: 5, Increment: 0.05}}},
+					"2",
+					nil,
+				},
+				{
+					"Very large precision value: expect error and an empty string in return",
+					openrtb_ext.PriceGranularity{Precision: int(^uint(0) >> 1), Ranges: []openrtb_ext.GranularityRange{{Max: 5, Increment: 0.05}}},
```

Comment:
Do we really want to test with about the largest possible integer when the limit is only 4? I would worry that this could potentially generate an error due to being silly large rather than violating our stated limit of 4. Would probably  generate a different error message, or error out in another way, so would still be flagged, but doesn't make sense as the number doesn't really compare to 4.

Reply:
You are totally right, that was too much. The latest version limits the precision value to 2^15 that both not panics and seems to not overflow the return string. I believe 2^15 = 32768 is more than enough. but let me know your take on this.